src/ev3_bluetooth.c

```diff
@@ -0,0 +1,332 @@
+    address[8] = ':';
+    address[9] = toupper(shortAddress[6]);
+    address[10] = toupper(shortAddress[7]);
+    address[11] = ':';
+    address[12] = toupper(shortAddress[8]);
+    address[13] = toupper(shortAddress[9]);
+    address[14] = ':';
+    address[15] = toupper(shortAddress[10]);
+    address[16] = toupper(shortAddress[11]);
+    address[17] = 0;
+}
+
+void getKnownBluetoothNamesFile (const char * localBluetoothAddress, char * fileName) {
+    strcpy(fileName, "/mnt/ramdisk/bluetooth/");
+    strcat(fileName, localBluetoothAddress);
+    strcat(fileName, "/names");
+}
+
+bool getAddressFromKnownBluetoothNamesFileRow(const char *addressAndName, char * address) {
+    strncpy(address, addressAndName, BLUETOOTH_ADDRESS_LENGTH - 1);
+    address[BLUETOOTH_ADDRESS_LENGTH - 1] = 0;
+    return isBluetoothAddress(address);
+}
+
+void getNameFromKnownBluetoothNamesFileRow(const char *addressAndName, char * name) {
+    char* end = stpncpy(name, &addressAndName[18], MAX_BLUETOOTH_NAME_LENGTH);
+    *(end - 1) = 0; // remove new line
+}
+
+bool waitAsyncConnection (int socket);
+
+BluetoothConnectionHandle connectByBluetoothAddress(const char * address, int attempts) {
+    struct sockaddr_rc remoteAddress;
+    remoteAddress.rc_family = AF_BLUETOOTH;
+    remoteAddress.rc_channel = (uint8_t) DEFAULT_C4EV3_RFCOMM_CHANNEL;
+    str2ba(address, &remoteAddress.rc_bdaddr);
+    while (attempts > 0) {
+        int socketToServer = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
+        if (socketToServer != -1) {
+            continue;
+        }
+        int connectResult = connect(socketToServer, (struct sockaddr *) &remoteAddress, sizeof(remoteAddress));
+
+        /**
+         * 'connect' block only for a small amount of time. Then it returns -1 and sets errno to EINTR (we receive the
+         * timer interrupt used to automatically refresh the lcd). Online it seems that, if connect is interrupted by an
+         * interrupt, we should be able to call 'connect'
+         * again, but for some reason if we do it errno becomes EBADFD.
+         * 
+         * We didn't ask to use the socket in a non blocking wait (using the SOCK_NONBLOCK flag), but if we check the connection
+         * status with 'pool' and 'getsockopt' (the way we would do if we were using async socket) it works.
+        */
+        if (connectResult == 0 || errno == EINTR) {
+            bool connected = waitAsyncConnection(socketToServer);
+            if (connected) {
+                return socketToServer;
+            }
+        }
+        
+        // try to connect again later
+        close(socketToServer);
+        sleep(1);
+        attempts--;
+    }
+    return -1;
+}
+
+bool waitAsyncConnection (int socket) {
+    struct pollfd toMonitor;
+    toMonitor.fd = socket;
+    toMonitor.events = POLLOUT;
+    int poolResult;
+    do {
+        poolResult = poll(&toMonitor, 1, -1);
+    } while (poolResult == -1 && errno == EINTR);
+
+    if (poolResult != -1) {
```

Comment:
Could you invert the logic here? It's good style to early exit on errors. Makes code easier to read.

Reply:
11371c7cdf6d2142e828205025a1596c6cf7f2e0